packages/target-tauri/translationfn/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+use std::collections::HashMap;
+
+use intl_pluralrules::{PluralCategory, PluralRuleType, PluralRules};
+use log::error;
+use regex::{self, Regex};
+use unic_langid::{LanguageIdentifier, LanguageIdentifierError};
+
+pub enum Substitution<'a> {
+    None,
+    String(Vec<&'a str>),
+    Quantity(usize),
+    QuantityFloat(f32),
+}
+
+pub struct TranslationEngine {
+    messages: HashMap<String, HashMap<String, String>>,
+    plural_rules: PluralRules,
+    var_finder: Regex,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("language id malformed")]
+    LangIdMalformed,
+    #[error("parsing of lang id failed {0:?}")]
+    LangIdParsingFailed(LanguageIdentifierError),
+    #[error("Plural rule not found {0}")]
+    PluralRuleNotFound(&'static str),
+    #[error("Invalid RegEx Pattern")]
+    RegEx(#[from] regex::Error),
+}
+
+impl TranslationEngine {
+    pub fn new(
+        messages: HashMap<String, HashMap<String, String>>,
+        langid: &str,
+    ) -> Result<Self, Error> {
+        let langid: LanguageIdentifier = langid
+            .split("_")
+            .next()
+            .ok_or(Error::LangIdMalformed)?
+            .parse()
+            .map_err(Error::LangIdParsingFailed)?;
+        let pr = PluralRules::create(langid, PluralRuleType::CARDINAL)
+            .map_err(Error::PluralRuleNotFound)?;
+        let var_finder = Regex::new(r"%\d\$[\w\d]")?;
+
+        Ok(TranslationEngine {
+            messages,
+            plural_rules: pr,
+            var_finder,
+        })
+    }
+    pub fn translate(&self, key: &str, substitution: Substitution) -> String {
+        if let Some(entry) = self.messages.get(key) {
+            match substitution {
+                Substitution::None => {
+                    if let Some(message) = entry.get("message") {
+                        message.to_owned()
+                    } else {
+                        error!("Message not existing for {key}");
+                        key.to_owned()
+                    }
+                }
+                Substitution::String(items) => {
+                    //TODO: remove unwrap, use log instead
```

Comment:
Shouldn't this be addressed prior to the merge?
We wouldn't want to panic in a translation function.